DataFormats/Provenance/interface/IndexIntoFile.h

```diff
@@ -969,6 +1107,9 @@ namespace edm {
     /// works in cases where those corrections are not needed.
     void sortVector_Run_Or_Lumi_Entries();
 
+    /// Run this check just after sorting
+    void checkForMissingRunOrLumiEntry() const;
```

Comment:
This function gets called every time a new IndexIntoFile is created. It is called once, not on every addEntry call.

The main purpose is to check for bugs in the source. If the source does not provide a complete and required sequence of Runs, Lumis, and Events that indirectly results in a proper sequence of calls to addEntry, the check can fail. I did not verify this, but I think an exception that causes a process to end early can also cause this check to fail. This is important because if we did not check we could create files that are not readable. We could hit assert failures in the Framework when a subsequent processing step tried to read a file.

That said, the check should never fail if everything was working OK. If profiling showed it was taking significant time, we might test our sources and decide they are working OK and remove this or only run it in limited testing contexts. My guess is that it is negligible, but I did not profile it.

Before this pull request and before concurrent runs, we did even more stringent tests in addEntry. We were testing that we received a very constrained sequence in calls to addEntry. But we cannot do that anymore because a call to addEntry for a Run could in principal come at almost any time. It will no longer be the case that we get events from a run and expect to get addEntry for the run before we get events from some other runs. I had to remove the more stringent checks from addEntry. And I replaced that with the less stringent requirement in this function. We still must get a call to addEntry for the run at some point, but it could come much later than it came before concurrent runs were allowed.